aiida/storage/sqlite_dos/backend.py

```diff
@@ -0,0 +1,187 @@
+###########################################################################
+"""Storage implementation using Sqlite database and disk-objectstore container."""
+from __future__ import annotations
+
+from functools import cached_property
+from pathlib import Path
+from tempfile import mkdtemp
+from typing import TYPE_CHECKING
+
+from disk_objectstore import Container
+from sqlalchemy import insert
+from sqlalchemy.orm import scoped_session, sessionmaker
+
+from aiida.manage import Profile
+from aiida.orm.implementation import BackendEntity
+from aiida.storage.psql_dos.models.settings import DbSetting
+from aiida.storage.sqlite_zip import models, orm
+from aiida.storage.sqlite_zip.migrator import get_schema_version_head
+from aiida.storage.sqlite_zip.utils import create_sqla_engine
+
+from ..psql_dos import PsqlDosBackend
+from ..psql_dos.migrator import REPOSITORY_UUID_KEY, PsqlDosMigrator
+
+if TYPE_CHECKING:
+    from aiida.repository.backend import DiskObjectStoreRepositoryBackend
+
+__all__ = ('SqliteDosStorage',)
+
+
+class SqliteDosMigrator(PsqlDosMigrator):
+    """Storage implementation using Sqlite database and disk-objectstore container."""
+
+    def __init__(self, profile: Profile) -> None:
+        # pylint: disable=super-init-not-called
+        filepath_database = Path(profile.storage_config['filepath']) / 'database.sqlite'
+        filepath_database.touch()
+
+        self.profile = profile
+        self._engine = create_sqla_engine(filepath_database)
+        self._connection = None
+
+    def get_container(self) -> Container:
+        """Return the disk-object store container.
+
+        :returns: The disk-object store container configured for the repository path of the current profile.
+        """
+        filepath_container = Path(self.profile.storage_config['filepath']) / 'container'
+        return Container(str(filepath_container))
+
+    def initialise_database(self) -> None:
+        """Initialise the database.
+
+        This assumes that the database has no schema whatsoever and so the initial schema is created directly from the
+        models at the current head version without migrating through all of them one by one.
+        """
+        models.SqliteBase.metadata.create_all(self._engine)
+
+        repository_uuid = self.get_repository_uuid()
+
+        # Create a "sync" between the database and repository, by saving its UUID in the settings table
+        # this allows us to validate inconsistencies between the two
+        self.connection.execute(
+            insert(DbSetting).values(key=REPOSITORY_UUID_KEY, val=repository_uuid, description='Repository UUID')
+        )
+
+        # finally, generate the version table, "stamping" it with the most recent revision
+        with self._migration_context() as context:
+            context.stamp(context.script, 'main@head')  # type: ignore[arg-type]
+            self.connection.commit()  # pylint: disable=no-member
+
+
+class SqliteDosStorage(PsqlDosBackend):
```

Comment:
I'm sceptical that here and in `SqliteDosMigrator` you can simply inherit from the PostgreSQL implementation and it will all just work, particularly migration, where you have differences in data types (like JSON and not JSONB).

You can also see methods like `delete_nodes_and_connections` which specifically uses `psql_dos` ORM models

Obviously testing would be required